test/micro/org/openjdk/bench/java/lang/reflect/Clazz.java

```diff
@@ -52,6 +55,34 @@ public Constructor getConstructor() throws NoSuchMethodException {
         return Clazz.class.getConstructor();
     }
 
+    /**
+     * Get modifiers for this class through reflection
+     *
+     * @return
+     * @throws NoSuchMethodException
+     */
+    @Benchmark
+    public int getModifiers() {
+        return Clazz.class.getModifiers();
+    }
+
+    Clazz[] clazzArray = new Clazz[1];
+    @Benchmark
+    public int getAppArrayModifiers() {
+        return clazzArray.getClass().getModifiers();
+    }
```

Comment:
I'm guessing this is the benchmarks that shows an extra load.  How about adding a benchmark that makes the Clazz[] final or @Stable, and see if that makes the extra load go away?

Reply:
Name                      Cnt  Base   Error   Test   Error  Unit  Change
getAppArrayModifiers       30 0.923 ± 0.004  1.260 ± 0.001 ns/op   0.73x (p = 0.000*)
getAppArrayModifiersFinal  30 0.922 ± 0.000  1.260 ± 0.001 ns/op   0.73x (p = 0.000*)

No it doesn't really help.  There's still an extra load.

Reply:
OK, if the extra load turns out to be a problem in the future, we could look into why the compilers are generating the load when the Class is known/constant.  If the old intrinsic was able to pull the constant out of the Klass, then surely we can do the same and pull the value from the Class field.

Reply:
Does static final help here?

Reply:
Yes.  Yes it does.

Reply:
Cases when a class mirror is a compile-time constant are already well-optimized. Non constant cases are the ones where missing optimization opportunities arise. 

In this particular case, C2 doesn't benefit from the observation that `Clazz[]` is a leaf type at runtime (no subclasses). Hence, a value loaded from a field typed as `Clazz[]` has exactly the same type and `clazzArray.getClass()` can be constant folded to `Clazz[].class`.  Rather than a common case, it feels more like a corner case. So, worth addressing as a follow-up enhancement.

Another scenario is a meet of 2 primitive array types (ends up as `bottom[]` in C2 type system), but I believe it hasn't been optimized before.